ouroboros-consensus-cardano-tools/src/Cardano/Tools/ImmDBServer/MiniProtocols.hs

```diff
@@ -0,0 +1,229 @@
+          where
+            miniProtocolLimits = limits N2N.defaultMiniProtocolParameters
+            miniProtocolRun    = ResponderProtocolOnly proto
+
+chainSyncServer ::
+     forall m blk a. (IOLike m, HasHeader blk)
+  => ImmutableDB m blk
+  -> BlockComponent blk a
+  -> ResourceRegistry m
+  -> ChainSyncServer a (Point blk) (Tip blk) m ()
+chainSyncServer immDB blockComponent registry = ChainSyncServer $ do
+    iterator <- ImmutableDB.streamAll immDB registry blockComponent
+    runChainSyncServer $ idle iterator Nothing
+  where
+    idle ::
+         Iterator m blk a
+      -> Maybe (Point blk) -- just-negotiated intersection, if any
+      -> ChainSyncServer a (Point blk) (Tip blk) m ()
+    idle iterator mIntersectionPt =
+        ChainSyncServer $ pure ServerStIdle {
+            recvMsgRequestNext   = handleRequestNext'
+          , recvMsgFindIntersect = handleFindIntersect iterator
+          , recvMsgDoneClient    = pure ()
+          }
+      where
+        handleRequestNext' = case mIntersectionPt of
+          Just pt -> do
+            tip <- getImmutableTip
+            pure $ Left $ SendMsgRollBackward pt tip $ idle iterator Nothing
+          Nothing -> handleRequestNext iterator
+
+    handleRequestNext ::
+         Iterator m blk a
+      -> m (Either (ServerStNext a (Point blk) (Tip blk) m ())
+                (m (ServerStNext a (Point blk) (Tip blk) m ())))
+    handleRequestNext iterator =
+       ImmutableDB.iteratorNext iterator >>= \case
+         ImmutableDB.IteratorExhausted ->
+           throwIO ReachedImmutableTip
+         ImmutableDB.IteratorResult a  -> Left <$> do
+           tip <- getImmutableTip
+           pure $ SendMsgRollForward a tip $ idle iterator Nothing
+
+    handleFindIntersect ::
+         Iterator m blk a
+      -> [Point blk]
+      -> m (ServerStIntersect a (Point blk) (Tip blk) m ())
+    handleFindIntersect iterator = go
+      where
+        go (pt : pts) =
+          ImmutableDB.streamAfterPoint immDB registry blockComponent pt >>= \case
+            Left _          -> go pts
+            Right iterator' -> do
+              ImmutableDB.iteratorClose iterator
+              tip <- getImmutableTip
+              pure $ SendMsgIntersectFound pt tip $ idle iterator' (Just pt)
+        go [] = do
+          tip <- getImmutableTip
+          pure $ SendMsgIntersectNotFound tip $ idle iterator Nothing
+
+    getImmutableTip :: m (Tip blk)
+    getImmutableTip = atomically (ImmutableDB.getTip immDB) <&> \case
+        Origin                         -> TipGenesis
+        NotOrigin ImmutableDB.Tip {..} -> Tip tipSlotNo tipHash tipBlockNo
+
+blockFetchServer ::
+     forall m blk a. (IOLike m)
+  => ImmutableDB m blk
+  -> BlockComponent blk a
+  -> ResourceRegistry m
+  -> BlockFetchServer a (Point blk) m ()
+blockFetchServer immDB blockComponent registry = idle
+  where
+    idle :: BlockFetchServer a (Point blk) m ()
+    idle = flip BlockFetchServer () \(ChainRange from to) -> do
```

Comment:
Smells to me like this `do` block could live upstream and be shared with the real BlockFetch server. Is my nose being overly optimistic? (Future work?)

Reply:
Yeah, the `blockFetchServer` logic could be abstracted out; it would need a mini-abstraction over `ChainDB.stream` vs `ImmutableDB.stream`. What about trying it in this PR, but postponing it for future work if I hit any roadblock?